### ISSUE TITLE
[BUG] React, Zustand, typing errors

### ISSUE THREAD

**Issue by ninest** [liveblocks/liveblocks#381]
**Describe the bug**

I am unable to get typing properly with `@liveblocks/zustand`.

**To Reproduce**

Stackblitz: https://stackblitz.com/edit/vitejs-vite-pwjssw?file=index.html,package.json,src%2Fstore.ts&terminal=dev

```tsx
import { createClient } from '@liveblocks/client';
import { middleware } from '@liveblocks/zustand';
import create from 'zustand';

export const client = createClient({
  publicApiKey: 'pk_example',
});

interface State {
  name: string;
  setName: (name: string) => void;
}

export const useStore = create<State>()(
  middleware(
    (set, get) => ({
      name: '',
      setName: (name: string) => {
        set({ ...get(), name });
      },
    }),
    {
      client,
    }
  )
);
```

<details>
<summary>Full error</summary>

```ts
(alias) middleware<ZustandState, JsonObject, LsonObject, BaseUserMeta, Json>(config: StateCreator<ZustandState, (partial: ZustandState | Partial<...> | ((state: ZustandState) => ZustandState | Partial<...>), replace?: boolean | undefined) => void, GetState<...>, StoreApi<...>>, options: Options<...>): StateCreator<...>
import middleware
Argument of type 'StateCreator<LiveblocksState<ZustandState, JsonObject, LsonObject, BaseUserMeta, Json>, (partial: LiveblocksState<ZustandState, JsonObject, LsonObject, BaseUserMeta, Json> | Partial<...> | ((state: LiveblocksState<...>) => LiveblocksState<...> | Partial<...>), replace?: boolean | undefined) => void, GetState<...>, S...' is not assignable to parameter of type 'StateCreator<SS, [], [never, unknown][], SS>'.
  Type 'StateCreator<LiveblocksState<ZustandState, JsonObject, LsonObject, BaseUserMeta, Json>, (partial: LiveblocksState<ZustandState, JsonObject, LsonObject, BaseUserMeta, Json> | Partial<...> | ((state: LiveblocksState<...>) => LiveblocksState<...> | Partial<...>), replace?: boolean | undefined) => void, GetState<...>, S...' is not assignable to type '(setState: (partial: SS | Partial<SS> | ((state: SS) => SS | Partial<SS>), replace?: boolean | undefined) => void, getState: GetState<SS>, store: StoreApi<...>, $$storeMutations: []) => SS'.
    Types of parameters 'setState' and 'setState' are incompatible.
      Type '(partial: SS | Partial<SS> | ((state: SS) => SS | Partial<SS>), replace?: boolean | undefined) => void' is not assignable to type 'never'.
```

</details>

**Expected behavior**

Types work out of the box

**Environment (please complete the following information):**

- `"@liveblocks/client": "^0.17.4"`
- `"@liveblocks/zustand": "^0.17.4"`

**Additional context**

I was able to fix the problem by doing something like this:

First I had to make some types:

```tsx
import { StoreApi, UseBoundStore } from "zustand";

type LiveblocksState = {
  liveblocks: { enterRoom: Function; leaveRoom: Function; others: any };
};
export type LiveblocksStore<T extends object> = UseBoundStore<
  StoreApi<T & LiveblocksState>
>;
```

Then create a Zustand store using slices with a few `ts-ignore`s:

```ts
import { middleware } from "@liveblocks/zustand";
import create, { StateCreator } from "zustand";
import { client } from "../liveblocks";
import { LiveblocksStore } from "../types/zustand";

interface State {
  name: null | string;
  setName: (name: string) => void;
}
const createStateSlice: StateCreator<State, [], []> = (set, get) => ({
  name: null,
  setName: (name) => {
    set({ ...get(), name });
  },
});

export const useStore: LiveblocksStore<State> = create<State>()(
  // @ts-ignore
  middleware(
    // @ts-ignore
    (...a) => ({
      // @ts-ignore
      ...createStateSlice(...a),
    }),
    {
      client,
      presenceMapping: {
        name: true,
      },
    }
  )
);

```

I'm using the slices to keep types intact in the store (for example, while writing the definition of the `setName` function).

**Comment by vinodsantharam** [comment]
Hi @ninest! Thanks for reporting this issue. We are currently working on it. The fix and an example should be released in the next 24 hours. Let us know if you still facing issues.

**Comment by ninest** [comment]
Thanks! I should also mention that I am using Vite and I hadn't seen #97. I did end up dropping `@liveblocks/zustand` and only using the React library, but I still had to add this to my `vite.config.js`:

```ts
export default defineConfig({
  optimizeDeps: {
    exclude: ["@liveblocks/react", "@liveblocks/client"],
  },
});
```

**Comment by GuillaumeSalles** [comment]
Hello @ninest,

Even if you're not using `zustand` anymore, just to let you know that we released [`0.17.7`](https://github.com/liveblocks/liveblocks/releases/tag/v0.17.7) to simplify the integration with typescript. It also works with [slices](https://github.com/liveblocks/liveblocks/blob/main/packages/liveblocks-zustand/types/storeWithSlices/useStore.ts). 

Note that it only works with `zustand` v3. v4 introduces breaking changes for middlewares so we'll probably release a `@liveblocks/zustand` v0.18 or v0.19 to support it.

Closing the issue but don't hesitate if you have any questions/comments.
